File: code/sorting/Sleep_Sort/Sleep_Sort.py
# Part of Cosmos by OpenGenus Foundation 
import threading
import time

_lk = threading.Lock()

class SleepSortThread(threading.Thread):
    def __init__(self, val):
        self.val = val
        threading.Thread.__init__(self)

    def run(self):
        global _lk
        # Thread is made to sleep in proportion to value
        time.sleep(self.val * 0.1)
        # Acquire lock
        _lk.acquire()
        print(self.val, end = " ")
        # Release lock
        _lk.release()

def SleepSort(list):
    ts = []

    # Intialize a thread corresponding to each element in list
    for i in list:
        t = SleepSortThread(i)
        ts.append(t)

    # Start all Threads
    for i in ts:
        i.start()

    # Wait for all threads to terminate
    for i in ts:
        i.join()

x = [2, 4, 3, 1, 6, 8, 4]
x = SleepSort(x)

File: code/sorting/src/sleep_sort/sleep_sort.scala
object Main
{
  def sleep_sort(xs: List[Int]): List[Int] =
  {
    var r: List[Int] = Nil
    var threads: List[Thread] = Nil
    for (x <- xs)
    {
      val t = new Thread(new Runnable
      {
        def run()
        {
          Thread.sleep(x)
          this.synchronized
          {
            r = r:+ x
          }
        }
      })
      threads = t :: threads
      t.start
    }

    for (t <- threads)
    {
      t.join()
    }
    r
  }
  def main(args: Array[String]) =
  {
    println(sleep_sort(List(100,99,3,15,2)))
  }
}
